File: packages/wasmvm/wasmlib/src/context.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// encapsulates standard host entities into a simple interface

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// smart contract interface with mutable access to state
#[derive(Clone, Copy)]
pub struct ScFuncContext {}

// reuse shared part of interface
impl ScSandbox for ScFuncContext {}

impl ScSandboxFunc for ScFuncContext {}

impl ScFuncClientContext for ScFuncContext {}

impl ScViewClientContext for ScFuncContext {
    fn client_contract(&self, h_contract: ScHname) -> ScHname {
        h_contract
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// smart contract view interface which has only immutable access to state
#[derive(Clone, Copy)]
pub struct ScViewContext {}

// reuse shared part of interface
impl ScSandbox for ScViewContext {}

impl ScSandboxView for ScViewContext {}

impl ScViewClientContext for ScViewContext {
    fn client_contract(&self, h_contract: ScHname) -> ScHname {
        h_contract
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/context.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

// encapsulates standard host entities into a simple interface

import {ScFuncCallContext, ScViewCallContext} from "./contract";
import {ScSandboxFunc, ScSandboxView} from "./sandbox";
import {ScChainID} from "./wasmtypes/scchainid";
import {ScHname} from "./wasmtypes/schname";

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// smart contract interface with mutable access to state
export class ScFuncContext extends ScSandboxFunc implements ScFuncCallContext {
    // host(): ScHost | null {
    //     return null;
    // }

    chainID(): ScChainID {
        return super.currentChainID();
    }

    initFuncCallContext(): void {
    }

    initViewCallContext(hContract: ScHname): ScHname {
        return hContract;
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// smart contract view interface which has only immutable access to state
export class ScViewContext extends ScSandboxView implements ScViewCallContext {

    chainID(): ScChainID {
        return super.currentChainID();
    }

    initViewCallContext(hContract: ScHname): ScHname {
        return hContract;
    }
}
